### ISSUE TITLE
Allow prioritizing certain dynamic templates when using ignore_dynamic_beyond_limit

### ISSUE THREAD

**Issue by flash1293** [elastic/elasticsearch#113275]
### Description

The `ecs@mappings` template is very helpful to avoid having to specify ecs field mappings manually. However, as it's using dynamic templates, it's possible known ECS fields won't be mapped if documents contain too many fields and `ignore_dynamic_beyond_limit` is set to true:

```
# logs-myoverflow-default doesn't exist yet, will use the default logs template which uses ecs@mappings
POST logs-myoverflow-default/_doc
{
  "field1": "...",
  "field2": "...",
  "field3": "...",
  // ...
  "field2000": "...",
  "host.name": "my-host",
}
```

As the ordering of fields in the incoming document is not respected, the fields that will make it into the mapping are not specified. In an unlucky case , very important fields like`host.name` won't be mapped and are not searchable.

While this situation is a misconfiguration of some kind (either the field limit is set to low or the shipper is shipping too many fields), the failure mode can be improved by prioritizing certain dynamic templates over others. In this specific example, the long list of field1,2,3 is matched by the `all_strings_to_keywords` template which doesn't specify any path match and is meant as a "catch all" fallback. `host.name` would be matched by `ecs_path_match_keyword_and_match_only_text` which is specified in a more specific way.

There are probably different ways to approach the problem and make it more deterministic which fields will "win out", I think the first one makes sense:

### Exhaust templates in order

Instead of taking in fields in a random order, try to exhaust the first defined template first by finding all fields that match it, then moving on to the next template instead of doing this on a per-field order. If the field limit is hit, the fields matching the templates defined first will have made it into the mapping, which seems like a natural behavior to me.

### Consider fields ordered alphabetically

Going through fields in alphabetic order will make things more deterministic, but doesn't give the user much control over which fields get mapped - I don't think it's a good solution.

### Allow specifying an explicit priority

Instead of using the inherent order of templates in the template, allow specifying a priority to guide the order of which templates to exhaust first. This seems harder to understand than just using the inherent array order.

### Prioritize more specific rules

Similar to CSS, come up with a specificity order to define which template is applied with priority. This seems more complex than necessary

cc @dakrone

**Comment by gwbrown** [comment]
I thiiiink I tagged this with the right `Mappings` label (we have two of them), please correct me if not.

**Comment by felixbarny** [comment]
An issue that wouldn't be solved with the proposed approaches is if there's not a single document with a lot of unique field but if the mapping explosion is a combination of different documents adding lots of fields. But maybe that's not as big of an issue because we'd expect the common fields to already be mapped before the field limit is hit.

As the fields are added in order they're applied in the source document, would it be an option to change the shipper so that the less important fields are added last? Having said that, I think it doesn't work if there's an ingest pipeline in the mix, as that re-writes the _source in a way that doesn't preserve ordering.

**Comment by flash1293** [comment]
> But maybe that's not as big of an issue because we'd expect the common fields to already be mapped before the field limit is hit.

Yeah, I was kind of leaning on this - I think with dynamic templates this problem is not really avoidable, but we can make the behavior a bit more consistent. The approach with exhausting templates in order is what I would expect to happen as a user as JSON doesn't guarantee an order of keys in an object.

> would it be an option to change the shipper so that the less important fields are added last? Having said that, I think it doesn't work if there's an ingest pipeline in the mix, as that re-writes the _source in a way that doesn't preserve ordering.

I don't think we should rely on ordering of keys in JSON objects - while some JSON parsers and serializers preserve order, others don't and it will never be a strong guarantee for anything. You mention ingest pipelines, but even for some shippers it might be impossible to actually define the order of keys if the used JSON library doesn't support this kind of thing.

**Comment by felixbarny** [comment]
> The approach with exhausting templates in order is what I would expect to happen as a user as JSON doesn't guarantee an order of keys in an object.

The tradeoff with that approach is that it would make document parsing a lot more inefficient. The way it's implemented atm is that ES is using a stream parser, to go through the source document without having to fully serialize it into a map (unlike what ingest processors are doing). So it goes through the fields as they're defined in the JSON source and looks up the dynamic template based on the path and the type of the current value.

**Comment by flash1293** [comment]
Good callout - this is definitely a significant downside of the proposed change, even if it's easier to conceptualize for users. 

I guess it would be possible to keep the stream parser, make optimistic choices like we do it now and then do a second pass at the end to see whether anything went wrong (it's not expected this will happen a lot in the first place), but it definitely has a smell and is increasing complexity.

@leehinman how do you feel about these tradeoffs?

**Comment by felixbarny** [comment]
> I don't think we should rely on ordering of keys in JSON objects - while some JSON parsers and serializers preserve order, others don't and it will never be a strong guarantee for anything.

With OTel, we have more control over how the JSON gets serialized. We can make it so that we send resource attributes first, then attributes, then metrics.

> make optimistic choices like we do it now and then do a second pass over chosen templates at the end to see whether anything went wrong

The issue is that by that time, the damage has already been done. The document has been parsed, and fields have been ignored.

**Comment by lucabelluccini** [comment]
Reducing the field count limit without real control by users and ability for users to opt-out has been perceived as an unannounced breaking change.
This is blindly invalidating:
- Dashboards and visualizations
- Saved searches
- Exports

All of this might be un-noticed by users except if the change is subtle.
The change introduced CAN be subtle as every rollover of a backing index will have different indices being indexed and other ignored as soon as
